x/group/server/msg_server.go

```diff
@@ -0,0 +1,604 @@
+	window, err := gogotypes.DurationFromProto(&timeout)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "maxVotingWindow time conversion")
+	}
+	endTime, err := gogotypes.TimestampProto(ctx.BlockTime().Add(window))
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "end time conversion")
+	}
+
+	m := &group.Proposal{
+		GroupAccount:        req.GroupAccount,
+		Comment:             comment,
+		Proposers:           proposers,
+		SubmittedAt:         *blockTime,
+		GroupVersion:        g.Version,
+		GroupAccountVersion: account.Version,
+		Result:              group.ProposalResultUnfinalized,
+		Status:              group.ProposalStatusSubmitted,
+		ExecutorResult:      group.ProposalExecutorResultNotRun,
+		Timeout:             *endTime,
+		VoteState: group.Tally{
+			YesCount:     "0",
+			NoCount:      "0",
+			AbstainCount: "0",
+			VetoCount:    "0",
+		},
+	}
+	if err := m.SetMsgs(msgs); err != nil {
+		return nil, sdkerrors.Wrap(err, "create proposal")
+	}
+
+	id, err := s.proposalTable.Create(ctx, m)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "create proposal")
+	}
+
+	// TODO: add event?
+
+	return &group.MsgCreateProposalResponse{ProposalId: group.ProposalID(id)}, nil
+}
+
+func (s serverImpl) Vote(ctx types.Context, req *group.MsgVoteRequest) (*group.MsgVoteResponse, error) {
+	id := req.ProposalId
+	voters := req.Voters
+	choice := req.Choice
+	comment := req.Comment
+
+	if err := assertCommentSize(comment, s.maxCommentSize(ctx), "comment"); err != nil {
+		return nil, err
+	}
+	if len(voters) == 0 {
+		return nil, sdkerrors.Wrap(group.ErrEmpty, "voters")
+	}
+
+	blockTime, err := gogotypes.TimestampProto(ctx.BlockTime())
+	if err != nil {
+		return nil, err
+	}
+	proposal, err := s.getProposal(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if proposal.Status != group.ProposalStatusSubmitted {
+		return nil, sdkerrors.Wrap(group.ErrInvalid, "proposal not open")
+	}
+	votingPeriodEnd, err := gogotypes.TimestampFromProto(&proposal.Timeout)
+	if err != nil {
+		return nil, err
+	}
+	if votingPeriodEnd.Before(ctx.BlockTime()) || votingPeriodEnd.Equal(ctx.BlockTime()) {
+		return nil, sdkerrors.Wrap(group.ErrExpired, "voting period has ended already")
+	}
+	var accountInfo group.GroupAccountInfo
+
+	address, err := sdk.AccAddressFromBech32(proposal.GroupAccount)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group account")
+	}
+	if err := s.groupAccountTable.GetOne(ctx, address.Bytes(), &accountInfo); err != nil {
+		return nil, sdkerrors.Wrap(err, "load group account")
+	}
+	if proposal.GroupAccountVersion != accountInfo.Version {
+		return nil, sdkerrors.Wrap(group.ErrModified, "group account was modified")
+	}
+
+	electorate, err := s.getGroupInfo(ctx, accountInfo.GroupId)
+	if err != nil {
+		return nil, err
+	}
+	if electorate.Version != proposal.GroupVersion {
+		return nil, sdkerrors.Wrap(group.ErrModified, "group was modified")
+	}
+
+	// count and store votes
+	for _, voterAddr := range voters {
```

Comment:
This doesn't sound good, if we are considering `voters`, we might need to consider array of votes directly (so each voter can have different `choice`)

Reply:
see https://github.com/regen-network/regen-ledger/pull/154#discussion_r562493715